game-core/src/main/java/games/strategy/engine/framework/startup/ui/editors/DiceServerEditor.java

```diff
@@ -188,39 +222,41 @@ private void checkFieldsAndNotify() {
   }
 
   public boolean areFieldsValid() {
-    final String toAddressText = toAddress.getText();
     final boolean toValid =
         setLabelValid(
-            !toAddressText.isEmpty() && PlayerEmailValidation.isValid(toAddressText), toLabel);
+            !toAddress.getText().isEmpty() && PlayerEmailValidation.isValid(toAddress.getText()),
+            toLabel);
     final boolean ccValid =
-        setLabelValid(PlayerEmailValidation.isValid(ccAddress.getText()), ccLabel);
-    final boolean serverValid = validateComboBox(servers, serverLabel);
-    final boolean allValid = serverValid && toValid && ccValid;
+        setLabelValid(
+            !ccAddress.getText().isEmpty() && PlayerEmailValidation.isValid(ccAddress.getText()),
+            ccLabel);
+    final boolean allValid = toValid && ccValid;
     testDiceButton.setEnabled(allValid);
+    testDiceButton.setToolTipText(
+        allValid
+            ? "Send a verified dice roll test email"
+            : "First enter a valid 'to' and 'cc' email address");
     return allValid;
   }
 
   public void applyToGameProperties(final GameProperties properties) {
-    properties.set(IRemoteDiceServer.NAME, servers.getSelectedItem());
     properties.set(IRemoteDiceServer.GAME_NAME, gameId.getText());
     properties.set(IRemoteDiceServer.EMAIL_1, toAddress.getText());
     properties.set(IRemoteDiceServer.EMAIL_2, ccAddress.getText());
   }
 
   public void populateFromGameProperties(final GameProperties properties) {
-    servers.setSelectedItem(properties.get(IRemoteDiceServer.NAME));
     gameId.setText(properties.get(IRemoteDiceServer.GAME_NAME, ""));
     toAddress.setText(properties.get(IRemoteDiceServer.EMAIL_1, ""));
     ccAddress.setText(properties.get(IRemoteDiceServer.EMAIL_2, ""));
   }
 
   public IRemoteDiceServer newDiceServer() {
-    final String selectedName = (String) servers.getSelectedItem();
-    assert selectedName != null;
-    final PropertiesDiceRoller roller = diceRollersByDisplayName.get(selectedName);
-    roller.setGameId(gameId.getText());
-    roller.setToAddress(toAddress.getText());
-    roller.setCcAddress(ccAddress.getText());
-    return roller;
+    return PropertiesDiceRoller.builder()
+        .diceRollerUri(ClientSetting.diceRollerUri.getValueOrThrow())
```

Comment:
Well, the preference choice is largely for auto-completion when loading a save. A person could change their properties file to use a different dice roller, they could launch a game, change the server, roll dice, and then change the server back before sending and emailing the game.

#thinkingOutLoud

Ultimately we would want to consider some sort of HMAC'ing scheme to really verify dice rolling integrity. IE: dice roller uses a private key to concat to a seqeunce of dice rolls and produces an HMAC from this. We could then verify that given an HMAC and a sequence that a specific roller did roll those results. We'd need to augment the dice roll data though to also include a dice roll id, in which case we'd have a data set like:

```
<100, 1>
<101, 3>
<102, 5>
<103, 2>
<104, 6>
````
We'd then sequence it, like: 10011013102510321046, append the private salt value, and then HMAC it.

Reply:
I have thought about this for a long time in the past and all of my approaches had one of 2 issues:
1. If a battle turns out to be a bad decision, a user can just reload the savegame and try again until the same battle is won. Without any public notification of rolls (which is currently done via emails) there's no way to prevent re-loading.
2. We could of course make battles deterministic by using a pseudo-random algorithm and just use a server signed random value to initialize it. This way every roll could be verified. But the determinism can be abused as well by just not fighting the battles you're guaranteed to lose.